Lint that warns when an elided lifetime ends up being a named lifetime (`elided_named_lifetimes`)

GrigorenkoPV opened the pull request:
As suggested in https://github.com/rust-lang/rust/issues/48686#issuecomment-1817334575

Fixes #48686

GrigorenkoPV committed:
Implement `elided_named_lifetimes` lint

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10434614799/job/28897680801) [(plain)](https://github.com/rust-lang/rust/commit/381915479df86d4c715e1505596feee067267dcd/checks/28897680801/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.813 Building wheels for collected packages: reuse
#13 2.815   Building wheel for reuse (pyproject.toml): started
#13 3.063   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.064   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.065   Stored in directory: /tmp/pip-ephem-wheel-cache-dacio7j1/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.067 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.471 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.471 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.010 Collecting virtualenv
#13 4.010 Collecting virtualenv
#13 4.057   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.133      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 76.7 MB/s eta 0:00:00
#13 4.177 Collecting distlib<1,>=0.3.7
#13 4.184   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.191      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 91.2 MB/s eta 0:00:00
#13 4.223 Collecting platformdirs<5,>=3.9.1
#13 4.231   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.271 Collecting filelock<4,>=3.12.2
#13 4.278   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.359 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.546 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      176064 kB
DirectMap2M:     7163904 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5495 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1868: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1871: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 39.6 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sat Aug 17 19:18:11 UTC 2024
  network time: Sat, 17 Aug 2024 19:18:11 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

GrigorenkoPV committed:
Fix `elided_named_lifetimes` in code

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10435230653/job/28898984026) [(plain)](https://github.com/rust-lang/rust/commit/54ea7651bde427d32ae2047e6fc22bb8d04d38a0/checks/28898984026/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.849 Building wheels for collected packages: reuse
#13 2.850   Building wheel for reuse (pyproject.toml): started
#13 3.093   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.093   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.094   Stored in directory: /tmp/pip-ephem-wheel-cache-ozxd5zca/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.096 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.489 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.489 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.007 Collecting virtualenv
#13 4.007 Collecting virtualenv
#13 4.057   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.276      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 26.3 MB/s eta 0:00:00
#13 4.331 Collecting platformdirs<5,>=3.9.1
#13 4.337   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.358 Collecting distlib<1,>=0.3.7
#13 4.378      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 44.4 MB/s eta 0:00:00
#13 4.415 Collecting filelock<4,>=3.12.2
#13 4.422   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.422   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.505 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.691 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      196544 kB
DirectMap2M:     7143424 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/0f26ee4fd95a1c046582dfb18892f520788e2c2c/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-0f26ee4fd95a1c046582dfb18892f520788e2c2c-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5495 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1868: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1871: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 48.3 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sat Aug 17 21:14:27 UTC 2024
  network time: Sat, 17 Aug 2024 21:14:27 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

cjgillot commented on compiler/rustc_resolve/src/late.rs:
```diff
@@ -2039,6 +2063,43 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
+                        if !self.allow_elided_static {
+                            self.r.lint_buffer.buffer_lint(
+                                lint::builtin::ELIDED_NAMED_LIFETIMES,
+                                missing.id_if_not_fake_or(self.crate_node_id),
```
In what cases are the ids fake? In those cases, we should probably just not lint.

GrigorenkoPV replied:
```rust
pub fn get_mut(&'static self, x: &mut u8) -> &'_ mut u8
```
is ok, but
```rust
pub fn get_mut(&'static self, x: &mut u8) -> &mut u8
```
is not.

Any `MissingLifetime` nodes with `LifetimeKind::Ampersand` get overlooked when we emit buffered lints, for some reason. After staring at the code for a bit, it looked like they get an ad-hoc NodeId not tied to any source code, and not preserved anywhere, because such a lifetime eventually gets replaced with something else.

Not linting here would defeat the whole point of the lint.

GrigorenkoPV replied:
> Any `MissingLifetime` nodes with `LifetimeKind::Ampersand`

There seem to be more cases beside this one, because I remember getting some more ICEs about not emitting all buffered lints when I ran it on `tests/ui`, but I haven't looked into those yet.

cjgillot commented on compiler/rustc_resolve/src/late.rs:
```diff
@@ -1550,6 +1567,13 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
+        self.allow_elided_static = true;
+        self.visit_ty(ty);
+        self.allow_elided_static = false;
+    }
```
You do not need this method. `allow_elided_static` should be equivalent to the presence of `LifetimeRibKind::Elided(LifetimeRes::Static)` in the stack.

GrigorenkoPV replied:
What I am trying to do here is to allow not writing `'static` in cases like
```rust
const A: &[u8] = &[];
static B: &str = "hello";
impl Trait{
    const C: &T = ...;
}
```

but still warn about things like
```
warning: elided lifetime has a name
 --> ./test.rs:6:50
  |
6 |     pub fn get_mut(&'static self, x: &mut u8) -> &mut u8 {
  |                                                  ^ this elided lifetime gets resolved as `'static`
  |
```

GrigorenkoPV replied:
I guess I should've called it `do_not_warn_about_elided_statics`.

GrigorenkoPV replied:
Done c726445411bcef39af9b4df5f86b0727d86d8929

cjgillot replied:
I really mean that this method should not exist. If you want stack-like behaviour, you need to inspect `self.lifetime_ribs`. And if some information is missing, as a field to `LifetimeRibKind`.

GrigorenkoPV replied:
Fixed 30f5b78eba5f8d8e82b1f41368db9cc08f5fcada

cjgillot commented on compiler/rustc_hir/src/def.rs:
```diff
@@ -796,8 +796,8 @@ impl<Id> Res<Id> {
-        /// Id of the generic parameter that introduced it.
-        param: LocalDefId,
+        /// `LocalDefId` & `Ident` of the generic parameter that introduced it.
+        param: (LocalDefId, Ident),
```
Why?

GrigorenkoPV replied:
To be able to have the lifetime name in the message:
```
warning: elided lifetime has a name
 --> ./test.rs:6:49
  |
6 |     pub fn get_mut<'a>(&'a self, x: &mut u8) -> &mut u8 {
  |                    --                           ^ this elided lifetime gets resolved as `'a`
  |                    |
  |                    lifetime `'a` introduced here
```
Ident contains span and name (as a Symbol). We could get the span back with `tcx.source_span(local_def_id)` or however it's spelled. But not the name: `tcx.def_name(id)` requires `HirId`, not `DefId`, and `tcx.local_def_id_to_hir_id(..)` for some reason makes rustc hang and eat lots of RAM when I try to call it (infinite cycle?).

cjgillot replied:
`tcx.item_name` takes a `DefId` and won't need HIR.

GrigorenkoPV replied:
Oh, sorry, completely overlooked it.

Fixed in d645641862714d02b327e0d9d7d45888a1100e1a

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10444318707/job/28918705733) [(plain)](https://github.com/rust-lang/rust/commit/1aaa280434ad5e49554b53a1f6b548f06a9fa21b/checks/28918705733/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.737 Building wheels for collected packages: reuse
#13 2.738   Building wheel for reuse (pyproject.toml): started
#13 2.986   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.987   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.987   Stored in directory: /tmp/pip-ephem-wheel-cache-_bs2trf1/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.990 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.387 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.387 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.912 Collecting virtualenv
#13 3.912 Collecting virtualenv
#13 3.947   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.027      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 73.6 MB/s eta 0:00:00
#13 4.080 Collecting platformdirs<5,>=3.9.1
#13 4.084   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.118 Collecting filelock<4,>=3.12.2
#13 4.140 Collecting distlib<1,>=0.3.7
#13 4.143   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.150      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 90.4 MB/s eta 0:00:00
#13 4.150      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 90.4 MB/s eta 0:00:00
#13 4.230 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.417 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      227264 kB
DirectMap2M:    10258432 kB
DirectMap1G:     8388608 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6de928dce9472b864f4e5d590dd7aa2075cb7551/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6de928dce9472b864f4e5d590dd7aa2075cb7551-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5495 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1868: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1871: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Using cached pip-24.2-py3-none-any.whl (1.8 MB)
Installing collected packages: pip
  Attempting uninstall: pip
  Attempting uninstall: pip
    Found existing installation: pip 24.1
    Uninstalling pip-24.1:
      Successfully uninstalled pip-24.1
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sun Aug 18 22:47:55 UTC 2024
  network time: Sun, 18 Aug 2024 22:47:55 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10444450543/job/28919003581) [(plain)](https://github.com/rust-lang/rust/commit/adb79013107b9ca1bdff1ef9d7a88f0fd6684c17/checks/28919003581/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.843 Building wheels for collected packages: reuse
#13 2.845   Building wheel for reuse (pyproject.toml): started
#13 3.103   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.104   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.104   Stored in directory: /tmp/pip-ephem-wheel-cache-bpw210dg/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.107 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.522 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.522 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.089 Collecting virtualenv
#13 4.089 Collecting virtualenv
#13 4.135   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.211      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 77.1 MB/s eta 0:00:00
#13 4.253 Collecting distlib<1,>=0.3.7
#13 4.257   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.265      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 87.1 MB/s eta 0:00:00
#13 4.295 Collecting platformdirs<5,>=3.9.1
#13 4.299   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.334 Collecting filelock<4,>=3.12.2
#13 4.337   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.423 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.615 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      210880 kB
DirectMap2M:     7129088 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6de928dce9472b864f4e5d590dd7aa2075cb7551/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6de928dce9472b864f4e5d590dd7aa2075cb7551-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5495 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1868: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1871: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 72.4 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sun Aug 18 23:13:44 UTC 2024
  network time: Sun, 18 Aug 2024 23:13:44 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10444738660/job/28919638542) [(plain)](https://github.com/rust-lang/rust/commit/c8b72208b1c335eec8a4ca8a798ed2951bb9a128/checks/28919638542/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.037 Building wheels for collected packages: reuse
#13 3.038   Building wheel for reuse (pyproject.toml): started
#13 3.296   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.297   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.297   Stored in directory: /tmp/pip-ephem-wheel-cache-xahnu1q4/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.300 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.707 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.708 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.255 Collecting virtualenv
#13 4.255 Collecting virtualenv
#13 4.309   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.479      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 34.1 MB/s eta 0:00:00
#13 4.535 Collecting platformdirs<5,>=3.9.1
#13 4.542   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.565 Collecting distlib<1,>=0.3.7
#13 4.583      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 57.8 MB/s eta 0:00:00
#13 4.619 Collecting filelock<4,>=3.12.2
#13 4.627   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.627   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.709 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.900 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 5.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      194496 kB
DirectMap2M:     7145472 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6de928dce9472b864f4e5d590dd7aa2075cb7551/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6de928dce9472b864f4e5d590dd7aa2075cb7551-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5495 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1868: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1871: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 46.4 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Mon Aug 19 00:05:31 UTC 2024
  network time: Mon, 19 Aug 2024 00:05:31 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10498672355/job/29084058979) [(plain)](https://github.com/rust-lang/rust/commit/9208741cfbddcb134bb7529d7ec9b889e32a2b1c/checks/29084058979/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.117 Building wheels for collected packages: reuse
#13 3.118   Building wheel for reuse (pyproject.toml): started
#13 3.378   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.379   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.379   Stored in directory: /tmp/pip-ephem-wheel-cache-j0x6brsg/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.382 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.801 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.801 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.353 Collecting virtualenv
#13 4.353 Collecting virtualenv
#13 4.408   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.583      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 33.0 MB/s eta 0:00:00
#13 4.636 Collecting distlib<1,>=0.3.7
#13 4.643   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.656      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 43.7 MB/s eta 0:00:00
#13 4.697 Collecting filelock<4,>=3.12.2
#13 4.705   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.742 Collecting platformdirs<5,>=3.9.1
#13 4.750   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.832 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.022 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 5.1s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      208832 kB
DirectMap2M:     8179712 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.06s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6b678c57b63b3062fb97130b3617b82657f59c80/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6b678c57b63b3062fb97130b3617b82657f59c80-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5502 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1868: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1871: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 45.0 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Wed Aug 21 23:00:29 UTC 2024
  network time: Wed, 21 Aug 2024 23:00:29 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10517281514/job/29141090226) [(plain)](https://github.com/rust-lang/rust/commit/7c36714cc078819feb7ff83febe33b0d1b6197e5/checks/29141090226/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.846 Building wheels for collected packages: reuse
#13 2.847   Building wheel for reuse (pyproject.toml): started
#13 3.089   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.090   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.090   Stored in directory: /tmp/pip-ephem-wheel-cache-3lbhbu89/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.092 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.482 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.483 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.001 Collecting virtualenv
#13 4.001 Collecting virtualenv
#13 4.057   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.262      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 28.0 MB/s eta 0:00:00
#13 4.308 Collecting distlib<1,>=0.3.7
#13 4.315   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.325      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 56.6 MB/s eta 0:00:00
#13 4.357 Collecting platformdirs<5,>=3.9.1
#13 4.364   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.406 Collecting filelock<4,>=3.12.2
#13 4.413   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.494 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.686 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      190400 kB
DirectMap2M:     8198144 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/eff09483c67e6fc96c8098ba46dce476162754c5/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-eff09483c67e6fc96c8098ba46dce476162754c5-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5508 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1868: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:1871: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 47.1 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Thu Aug 22 23:48:13 UTC 2024
  network time: Thu, 22 Aug 2024 23:48:13 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

GrigorenkoPV commented:
> `rustc_errors::translation` was changed

- replaced some elided lifetimes with explicit ones

> Some changes occurred in src/tools/clippy

- replaced some elided lifetimes with explicit ones
- blessed some tests, because there is a new lint

> The Miri subtree was changed

- replaced some elided lifetimes with explicit ones

cjgillot commented:
Could you submit d214d0b0f2e221fc97dee444ca1686086568f876 as a standalone PR? Is can be merged while we review this PR.

cjgillot commented:
Thanks!

As for the matter with `crate_node_id`, it will prevent users from controlling lint behaviour using `allow`/`deny` attributes. The missing node-ids are those which correspond to a `ElidedAnchor` resolution.

2 alternative ideas:
- channel the correct NodeId for the lint to use;
- emit the lints when seeing the `ElidedAnchor` in `record_lifetime_res`.

GrigorenkoPV commented:
> As for the matter with `crate_node_id`, it will prevent users from controlling lint behaviour using `allow`/`deny` attributes. The missing node-ids are those which correspond to a `ElidedAnchor` resolution.

Ah, I was worried that it might cause some sort of a problem, and indeed it does. Oh well. I seem to have fixed it and I also added a test for this.

cjgillot commented on compiler/rustc_resolve/src/late.rs:
```diff
@@ -1916,6 +1928,7 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             };
             let missing_lifetime = MissingLifetime {
                 id: node_ids.start,
+                related_node_that_exists_in_source: finalize_node,
```
Why not segment_id?

GrigorenkoPV replied:
No good reason other than I have overlooked it. I will push a fix soon, and probably rebase as well. It's been quite some time, plus #129510 will probably land soon.

cjgillot commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -102,6 +102,13 @@ fn import_candidate_to_enum_paths(suggestion: &ImportSuggestion) -> (String, Str
+    /// in a sense that they are temporary and not get preserved down the line,
+    /// which means that the lints for those nodes will not get emitted.
+    /// To combat this, we can try to use some other `NodeId`s as a fallback option.
+    pub related_node_that_exists_in_source: NodeId,
```
Let's call this one id_for_lint, shall we?

GrigorenkoPV replied:
Fair enough

cjgillot commented on compiler/rustc_resolve/src/late.rs:
```diff
@@ -2039,8 +2052,44 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
+                            self.r.lint_buffer.buffer_lint(
+                                lint::builtin::ELIDED_NAMED_LIFETIMES,
+                                missing.related_node_that_exists_in_source,
+                                missing.span,
```
If you destructure a variable, please do it completely. Otherwise it gets confusing.

GrigorenkoPV replied:
Fair enough

GrigorenkoPV commented:
@rustbot authot

GrigorenkoPV commented:
@rustbot ready

cjgillot commented:
Thanks for bearing with me through this!
Do you mind adding a suggestion to this lint in a follow-up PR?
@bors r+

bors commented:
:pushpin: Commit 3a7d0ab1c48f763765195d57d761314325ae54c0 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 3a7d0ab1c48f763765195d57d761314325ae54c0 -->
<!-- homu: {"type":"Approved","sha":"3a7d0ab1c48f763765195d57d761314325ae54c0","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GrigorenkoPV commented:
> Thanks for bearing with me through this!

Thank *you* for your patience and guidance, I appreciate it!

> Do you mind adding a suggestion to this lint in a follow-up PR?

Yes (well, technically "No", I do *not* mind, heh), but a bit later as I'm going to be busy in the next couple of days.

matthiaskrgr commented:
@bors r-
https://github.com/rust-lang/rust/pull/129591#issuecomment-2309044568

saethlin commented:
@GrigorenkoPV The PR description makes it look like this PR is very incomplete. This PR description will become part of the commit history of the repo when the PR is merged. Perhaps you want to update the PR description before this merges?

bors commented:
:umbrella: The latest upstream changes (presumably #129750) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot commented:
With the same idea as @saethlin's comment, could you squash the commits? To remove some back-and-forth changes.
r=me then

GrigorenkoPV committed:
elided_named_lifetimes: bless & add tests

GrigorenkoPV commented:
Done!

I've squashed it into three commits:
1. Implementation
2. Fixing the remaining (after #129510) places in code where the lint would fire.
3. Adding & blessing tests.

cjgillot commented:
Great!
@bors r=

bors commented:
:pushpin: Commit a9b959a020cc156ef57a972c83381aabc43b1174 has been approved by ``

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r= a9b959a020cc156ef57a972c83381aabc43b1174 -->
<!-- homu: {"type":"Approved","sha":"a9b959a020cc156ef57a972c83381aabc43b1174","approver":"","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

cjgillot commented:
Typo.
@bors r-

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit a9b959a020cc156ef57a972c83381aabc43b1174 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot a9b959a020cc156ef57a972c83381aabc43b1174 -->
<!-- homu: {"type":"Approved","sha":"a9b959a020cc156ef57a972c83381aabc43b1174","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

kpreid commented:
This PR implements a lint named `elided_named_lifetimes`, but that name wasn't previously mentioned literally in any comment, so it doesn't show up in PR search. I hope this comment will fix that.

GrigorenkoPV commented:
True, thanks for pointing out! I've updated the title.

irbull commented:
Part of the Deno project is built with Rust +nightly and is being hit by this change-set with the following error:

```
error: elided lifetime has a name
  --> core/runtime/snapshot.rs:32:19
   |
32 | ) -> (V8Snapshot, SerializableSnapshotSidecarData) {
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this elided lifetime gets resolved as `'static`
```

The fix is easy [1], but I thought I would leave this here in case others are being hit by it and hopefully google can index "elided lifetime has a name".

[1] https://github.com/denoland/deno_core/pull/895

GrigorenkoPV commented:
> The fix is easy

There is an upcoming PR implementing suggestions for this lint https://github.com/rust-lang/rust/pull/129840. Currently it is in a conflict with some of clippy's suggestions, but maybe I can mark it non-machine-applicable for now, so that there are at least human-readable suggestions preventing confusion.

Thank you for your report!
